app/interactors/add_coupon.rb

```diff
@@ -2,18 +2,20 @@ class AddCoupon
   include Interactor
 
   def call
-    return context.fail! if coupon_attached?
-
-    return context.fail! unless coupon_valid?
+    return context.fail! unless coupon_can_be_attached?
 
     coupon = Coupon.find_by(code: context.code)
 
     context.current_order.update(coupon: coupon)
-    context.current_order.decrement!(:total_price, context.current_order.discount)
+    UpdateTotalPrice.call(context.current_order, coupon: coupon)
   end
 
   private
 
+  def coupon_can_be_attached?
+    !coupon_attached? && coupon_valid?
```

Comment:
It's hard to read and understand
`order_without_coupon? && coupon_valid?` looks better